Add PidFd::{kill, wait, try_wait}

the8472 opened the pull request:
#117957 changed `Child` kill/wait/try_wait to use its pidfd instead of the pid, when one is available.
This PR extracts those implementations and makes them available on `PidFd` directly.

The `PidFd` implementations differ significantly from the corresponding `Child` methods:

* the methods can be called after the child has been reaped, which will result in an error but will be safe. This state is not observable in `Child` unless something stole the zombie child
* the `ExitStatus` is not kept, meaning that only the first time a wait succeeds it will be returned
* `wait` does not close stdin
* `wait` only requires `&self` instead of `&mut self` since there is no state to maintain and subsequent calls are safe

Tracking issue: #82971

the8472 committed:
Add PidFd::{kill, wait, try_wait}

cuviper commented on library/std/src/os/linux/process.rs:
```diff
@@ -47,63 +48,98 @@ struct FileDesc;
+    ///
+    /// Unlike [`Child::wait`] it does not ensure that the stdin handle is closed.
+    /// Additionally it will not return an `ExitStatus` if the child
+    /// has alrady been reaped. Instead an error will be returned.
```
```suggestion
    /// has already been reaped. Instead an error will be returned.
```

cuviper commented:
If you call `child.pidfd().wait()` (or with `take_pidfd()`), doesn't that break `child.wait()`?

the8472 commented:
Good point. It shouldn't break it immediately since the pid won't be recycled until the fd is dropped (not confident, haven't tested this). But once one does drop it, yeah it'd become an issue.

But people do want to take out pidfds, e.g. for process management in tokio.
So either we have to make it `take_pidfd(self)` so it consumes the child or try_clone the fd instead so the child can hold onto its own copy.

the8472 committed:
to extract a pidfd we must consume the child

As long as a pidfd is on a child it can be safely reaped. Taking it
would mean the child would now have to be awaited through its pid, but could also
be awaited through the pidfd. This could then suffer from a recycling race.

the8472 commented:
I've replaced `take_pidfd` with `into_pidfd` which consumes the child.

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9507486112/job/26207048833) [(plain)](https://github.com/rust-lang/rust/commit/b5be21ddb1c8d6128274b102bc1618502d24ee3b/checks/26207048833/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
.......... (120/124)
...F       (124/124)


/checkout/tests/rustdoc-gui/help-page.goml help-page... FAILED
[ERROR] `tests/rustdoc-gui/utils.goml` line 16
    from `tests/rustdoc-gui/help-page.goml` line 23
    from `tests/rustdoc-gui/help-page.goml` line 32: Error: Evaluation failed: The following errors happened: [localStorage item "rustdoc-theme" (of value "ayu") != "null"]: for command `assert-local-storage: {"rustdoc-theme": |theme|}`
Error: ()
Build completed unsuccessfully in 0:03:29
  local time: Thu Jun 13 23:06:17 UTC 2024
  network time: Thu, 13 Jun 2024 23:06:17 GMT
```

</details>

cuviper commented:
OK, `into_pidfd()` is better to make sure we avoid pid reuse on the child, but there's still an issue with the saved status. That is, `child.pidfd().wait()` cannot save the status, so it's lost from `child.wait()`. (Ditto `try_wait()`)

At the very least, this should be an unresolved question on the tracking issue. Apart from that, `r=me`.

the8472 commented:
Yes it'll be lost but in those cases calling `Child::wait` will return an error as `PidFd::wait` would, not  an incorrect value.
This is indeed new behavior, but any code using PidFd will also be new.

I'll note it in the tracking issue.

@bors r=cuviper rollup

bors commented:
:pushpin: Commit 210400a1bea715c2b1463349d79aac661429519a has been approved by `cuviper`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=cuviper 210400a1bea715c2b1463349d79aac661429519a -->
<!-- homu: {"type":"Approved","sha":"210400a1bea715c2b1463349d79aac661429519a","approver":"cuviper","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

matthiaskrgr commented:
@bors r- https://github.com/rust-lang/rust/pull/126775#issuecomment-2182144955

workingjubilee commented:
@bors rollup=iffy

the8472 commented:
fixed the import, docs build for x86_64-pc-windows-gnu

@bors r=cuviper

bors commented:
:pushpin: Commit 8abf149bde08d8f1fc50517bbe4d50368740d054 has been approved by `cuviper`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=cuviper 8abf149bde08d8f1fc50517bbe4d50368740d054 -->
<!-- homu: {"type":"Approved","sha":"8abf149bde08d8f1fc50517bbe4d50368740d054","approver":"cuviper","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 8abf149bde08d8f1fc50517bbe4d50368740d054 with merge 06c6b72f2d9cf4b2d7df8107ed4fae5b383e5316...
<!-- homu: {"type":"BuildStarted","head_sha":"8abf149bde08d8f1fc50517bbe4d50368740d054","merge_sha":"06c6b72f2d9cf4b2d7df8107ed4fae5b383e5316"} -->

rust-log-analyzer commented:
The job **`x86_64-msvc`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/9621092085/job/26540411939) [(plain)](https://github.com/rust-lang-ci/rust/commit/06c6b72f2d9cf4b2d7df8107ed4fae5b383e5316/checks/26540411939/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
failures:

---- [ui] tests\ui\lto\lto-duplicate-symbols.rs stdout ----

error: auxiliary build of "C:\\a\\rust\\rust\\tests\\ui\\lto\\auxiliary\\lto-duplicate-symbols1.rs" failed to compile: 
status: exit code: 1
command: PATH="C:\a\rust\rust\build\x86_64-pc-windows-msvc\stage2\bin;C:\Program Files (x86)\Windows Kits\10\bin\x64;C:\Program Files (x86)\Windows Kits\10\bin\10.0.22621.0\x64;C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.40.33807\bin\HostX64\x64;C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.40.33807\bin\HostX64\x64;C:\a\rust\rust\build\x86_64-pc-windows-msvc\stage0-bootstrap-tools\x86_64-pc-windows-msvc\release\deps;C:\a\rust\rust\build\x86_64-pc-windows-msvc\stage0\bin;C:\Program Files\PowerShell\7;C:\a\_temp\msys64\mingw64\bin;C:\a\_temp\msys64\usr\local\bin;C:\a\_temp\msys64\usr\bin;C:\a\_temp\msys64\usr\bin;C:\a\rust\rust\ninja;C:\a\rust\rust\sccache;C:\a\_temp\setup-msys2;C:\Program Files\MongoDB\Server\5.0\bin;C:\aliyun-cli;C:\vcpkg;C:\Program Files (x86)\NSIS;C:\tools\zstd;C:\Program Files\Mercurial;C:\hostedtoolcache\windows\stack\2.15.7\x64;C:\cabal\bin;C:\ghcup\bin;C:\mingw64\bin;C:\Program Files\dotnet;C:\Program Files\MySQL\MySQL Server 8.0\bin;C:\Program Files\R\R-4.4.0\bin\x64;C:\SeleniumWebDrivers\GeckoDriver;C:\SeleniumWebDrivers\EdgeDriver;C:\SeleniumWebDrivers\ChromeDriver;C:\Program Files (x86)\sbt\bin;C:\Program Files (x86)\GitHub CLI;C:\Program Files\Git\bin;C:\Program Files (x86)\pipx_bin;C:\npm\prefix;C:\hostedtoolcache\windows\go\1.21.11\x64\bin;C:\hostedtoolcache\windows\Python\3.9.13\x64\Scripts;C:\hostedtoolcache\windows\Python\3.9.13\x64;C:\hostedtoolcache\windows\Ruby\3.0.7\x64\bin;C:\Program Files\OpenSSL\bin;C:\tools\kotlinc\bin;C:\hostedtoolcache\windows\Java_Temurin-Hotspot_jdk\8.0.412-8\x64\bin;C:\Program Files\ImageMagick-7.1.1-Q16-HDRI;C:\Program Files\Microsoft SDKs\Azure\CLI2\wbin;C:\ProgramData\kind;C:\ProgramData\docker-compose;C:\ProgramData\Chocolatey\bin;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0;C:\Windows\System32\OpenSSH;C:\Program Files\dotnet;C:\Program Files\PowerShell\7;C:\Program Files\Microsoft\Web Platform Installer;C:\Program Files\TortoiseSVN\bin;C:\Program Files\Microsoft SQL Server\Client SDK\ODBC\170\Tools\Binn;C:\Program Files\Microsoft SQL Server\150\Tools\Binn;C:\Program Files (x86)\Windows Kits\10\Windows Performance Toolkit;C:\Program Files (x86)\WiX Toolset v3.14\bin;C:\Program Files\Microsoft SQL Server\130\DTS\Binn;C:\Program Files\Microsoft SQL Server\140\DTS\Binn;C:\Program Files\Microsoft SQL Server\150\DTS\Binn;C:\Program Files\Microsoft SQL Server\160\DTS\Binn;C:\Strawberry\c\bin;C:\Strawberry\perl\site\bin;C:\Strawberry\perl\bin;C:\ProgramData\chocolatey\lib\pulumi\tools\Pulumi\bin;C:\Program Files\CMake\bin;C:\ProgramData\chocolatey\lib\maven\apache-maven-3.8.7\bin;C:\Program Files\Microsoft Service Fabric\bin\Fabric\Fabric.Code;C:\Program Files\Microsoft SDKs\Service Fabric\Tools\ServiceFabricLocalClusterManager;C:\Program Files\nodejs;C:\Program Files\Git\cmd;C:\Program Files\Git\mingw64\bin;C:\Program Files\Git\usr\bin;C:\Program Files\GitHub CLI;C:\tools\php;C:\Program Files (x86)\sbt\bin;C:\Program Files\Amazon\AWSCLIV2;C:\Program Files\Amazon\SessionManagerPlugin\bin;C:\Program Files\Amazon\AWSSAMCLI\bin;C:\Program Files\Microsoft SQL Server\130\Tools\Binn;C:\Program Files\LLVM\bin;C:\Users\runneradmin\.dotnet\tools;C:\Users\runneradmin\.cargo\bin;C:\Users\runneradmin\AppData\Local\Microsoft\WindowsApps;C:\a\_temp\msys64\usr\bin\site_perl;C:\a\_temp\msys64\usr\bin\vendor_perl;C:\a\_temp\msys64\usr\bin\core_perl" "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage2\\bin\\rustc.exe" "C:\\a\\rust\\rust\\tests\\ui\\lto\\auxiliary\\lto-duplicate-symbols1.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=C:\\Users\\runneradmin\\.cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=C:\\a\\rust\\rust\\vendor" "--sysroot" "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage2" "--target=x86_64-pc-windows-msvc" "--check-cfg" "cfg(FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--out-dir" "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\test\\ui\\lto\\lto-duplicate-symbols\\auxiliary" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\native\\rust-test-helpers" "-L" "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\test\\ui\\lto\\lto-duplicate-symbols\\auxiliary"
--- stderr -------------------------------
--- stderr -------------------------------
error: couldn't create a temp dir: Access is denied. (os error 5) at path "C:\\a\\_temp\\msys64\\tmp\\rustcMwb8Le"
error: aborting due to 1 previous error
------------------------------------------


---
test result: FAILED. 16875 passed; 1 failed; 250 ignored; 0 measured; 2 filtered out; finished in 379.97s

Some tests failed in compiletest suite=ui mode=ui host=x86_64-pc-windows-msvc target=x86_64-pc-windows-msvc
Build completed unsuccessfully in 0:40:48
make: *** [Makefile:102: ci-msvc-ps1] Error 1
  network time: Sat, 22 Jun 2024 00:38:20 GMT
##[error]Process completed with exit code 2.
Post job cleanup.
[command]"C:\Program Files\Git\bin\git.exe" version
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/9621092085/job/26541389036)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/9621092085/job/26541389036","builder_name":"checks-actions"} -->

cuviper commented:
@bors retry

bors commented:
:hourglass: Testing commit 8abf149bde08d8f1fc50517bbe4d50368740d054 with merge 10e1f5d2121831991a1bebb8cda044a1677aaddf...
<!-- homu: {"type":"BuildStarted","head_sha":"8abf149bde08d8f1fc50517bbe4d50368740d054","merge_sha":"10e1f5d2121831991a1bebb8cda044a1677aaddf"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/9622605705/job/26545617226)
Approved by: cuviper
Pushing 10e1f5d2121831991a1bebb8cda044a1677aaddf to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"cuviper","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/9622605705/job/26545617226"},"merge_sha":"10e1f5d2121831991a1bebb8cda044a1677aaddf"} -->

rust-timer commented:
Finished benchmarking commit (10e1f5d2121831991a1bebb8cda044a1677aaddf): [comparison URL](https://perf.rust-lang.org/compare.html?start=fcae62649e66b86eaf742414d8fc9c05d58e0a7e&end=10e1f5d2121831991a1bebb8cda044a1677aaddf&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=fcae62649e66b86eaf742414d8fc9c05d58e0a7e&end=10e1f5d2121831991a1bebb8cda044a1677aaddf&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.3% | [-0.3%, -0.3%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.3% | [-0.3%, -0.3%] | 1     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=fcae62649e66b86eaf742414d8fc9c05d58e0a7e&end=10e1f5d2121831991a1bebb8cda044a1677aaddf&stat=max-rss)
This benchmark run did not return any relevant results for this metric.

### [Cycles](https://perf.rust-lang.org/compare.html?start=fcae62649e66b86eaf742414d8fc9c05d58e0a7e&end=10e1f5d2121831991a1bebb8cda044a1677aaddf&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=fcae62649e66b86eaf742414d8fc9c05d58e0a7e&end=10e1f5d2121831991a1bebb8cda044a1677aaddf&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 693.147s -> 695.587s (0.35%)
**Artifact size**: 326.88 MiB -> 326.88 MiB (0.00%)
